internal/operator-controller/catalogmetadata/compare/compare.go

```diff
@@ -1,25 +1,71 @@
 package compare
 
 import (
+	"slices"
+	"strings"
+
+	mmsemver "github.com/Masterminds/semver/v3"
+	bsemver "github.com/blang/semver/v4"
 	"k8s.io/apimachinery/pkg/util/sets"
 
 	"github.com/operator-framework/operator-registry/alpha/declcfg"
 
 	"github.com/operator-framework/operator-controller/internal/operator-controller/bundleutil"
+	slicesutil "github.com/operator-framework/operator-controller/internal/shared/util/slices"
 )
 
-// ByVersion is a sort "less" function that orders bundles
-// in inverse version order (higher versions on top).
-func ByVersion(b1, b2 declcfg.Bundle) int {
-	v1, err1 := bundleutil.GetVersion(b1)
-	v2, err2 := bundleutil.GetVersion(b2)
-	if err1 != nil || err2 != nil {
-		return compareErrors(err1, err2)
+// NewVersionRange returns a function that tests whether a semver version is in the
+// provided versionRange. The versionRange provided to this function can be any valid semver
+// version string or any range that matches the syntax defined in the Masterminds semver library.
+//
+// When the provided version range is a valid semver version that includes build metadata, then the
+// returned function will only match an identical version with the same build metadata.
+//
+// When the provided version range is a valid semver version that does NOT include build metadata,
+// then the returned function will match any version that matches the semver version, ignoring the
+// build metadata of matched versions.
+//
+// Otherwise, Masterminds semver constraints semantics are used to match versions.
```

Comment:
should not both libraries implement semver spec? would it be possible to use just single library?

Reply:
The main reason we have both is that they have different implementations of version ranges. OLMv0 used blang's version ranges for `skipRange` semantics, so we need to continue using that for back-compat reasons.

However, blang's semver range semantics have some footguns and are not all that ergonomic. So we chose Masterminds range semantics for `spec.source.catalog.version`.

Reply:
Thanks, for the explanation, perhaps you could add to the comment in the code?

Reply:
> Thanks, for the explanation, perhaps you could add to the comment in the code?

We already have that documented here: https://github.com/operator-framework/operator-controller/blob/85e8cbfa106594ad25a00ceb597dec1b8eae2e9e/internal/operator-controller/catalogmetadata/filter/successors.go#L54-L58

Do you think that is sufficient?